naga/src/back/msl/writer.rs

```diff
@@ -675,6 +681,125 @@ impl<W: Write> Writer<W> {
+    /// This rewrite could occur even if the conditional assignment appears
+    /// *before* the `while` loop, as long as `i < 10` by the time the loop is
+    /// reached. This would allow the attacker to save the results of
+    /// unauthorized reads somewhere accessible before entering the infinite
+    /// loop. But even worse, the MSL compiler has been observed to simply
+    /// delete the infinite loop entirely, so that even code dominated by the
+    /// loop becomes reachable. This would make the attack even more flexible,
+    /// since shaders that would appear to never terminate would actually exit
+    /// nicely, after having stolen data from elsewhere in the GPU address
+    /// space.
+    ///
+    /// Ideally, Naga would prevent UB entirely via some means that persuades
+    /// the MSL compiler that no loop Naga generates is infinite. One approach
+    /// would be to add inline assembly to each loop that is annotated as
+    /// potentially branching out of the loop, but which in fact generates no
+    /// instructions. Unfortunately, inline assembly is not handled correctly by
+    /// some Metal device drivers. Further experimentation hasn't produced a
+    /// satisfactory approach.
+    ///
+    /// Instead, we accept that the MSL compiler may determine that some loops
+    /// are infinite, and focus instead on preventing the range analysis from
+    /// being affected. We transform *every* loop into something like this:
+    ///
+    /// ```ignore
+    /// if (volatile bool unpredictable = true; unpredictable)
+    ///     while (true) { }
+    /// ```
+    ///
+    /// Since the `volatile` qualifier prevents the compiler from assuming that
+    /// the `if` condition is true, it cannot be sure the infinite loop is
+    /// reached, and thus it cannot assume the entire structure is unreachable.
+    /// This prevents the range analysis impact described above.
+    ///
+    /// Unfortunately, what makes this a kludge, not a hack, is that this
+    /// solution leaves the GPU executing a pointless conditional branch, at
+    /// runtime, before each loop. There's no part of the system that has a
+    /// global enough view to be sure that `unpredictable` is true, and remove
+    /// it from the code.
+    ///
+    /// To make our output a bit more legible, we pull the condition out into a
+    /// preprocessor macro defined at the top of the module.
+    fn emit_loop_reachable_macro(&mut self) -> BackendResult {
+        if !self.loop_reachable_macro_name.is_empty() {
+            return Ok(());
+        }
+
+        self.loop_reachable_macro_name = self.namer.call("LOOP_IS_REACHABLE");
+        let loop_reachable_volatile_name = self.namer.call("unpredictable_jump_over_loop");
+        writeln!(
+            self.out,
+            "#define {} if (volatile bool {} = true; {})",
+            self.loop_reachable_macro_name,
+            loop_reachable_volatile_name,
+            loop_reachable_volatile_name,
+        )?;
```

Comment:
thought: IIRC, this approach is effectively cribbed from the Chromium WebGPU team's Dawn component (i.e., https://github.com/gfx-rs/wgpu/issues/4972#issuecomment-2326848790). It feels important to cite this as an influence in our implementation of this workaround, and not allow a reader to assume that this understanding is entirely our own (which is what I think a typical reader will do with the lengthy explanation above).

☝🏻This can be punted to a follow-up commit.

Reply:
Absolutely - I neglected to give credit where it was due.